src/main/java/com/google/cloud/dataproc/templates/main/DataProcTemplate.java

```diff
@@ -15,86 +15,116 @@
+   * Parse command line arguments
+   *
+   * @param args command line arguments
+   * @return parsed arguments
+   */
+  public static CommandLine parseArguments(String... args) {
+    Options options = new Options();
+
+    Option templateOption = new Option(TEMPLATE_NAME, "the name of the template to run");
+    templateOption.setRequired(true);
+    templateOption.setArgs(1);
+    options.addOption(templateOption);
+
+    @SuppressWarnings("AccessStaticViaInstance")
+    Option propertyOption =
+        OptionBuilder.withValueSeparator()
+            .hasArgs(2)
+            .withArgName("property=value")
+            .withLongOpt("prop")
+            .withDescription("Value for given property")
+            .create();
+    options.addOption(propertyOption);
+
+    CommandLineParser parser = new BasicParser();
+    LOGGER.info("Parsing arguments {}", (Object) args);
+    try {
+      return parser.parse(options, args, true);
+    } catch (ParseException e) {
+      throw new IllegalArgumentException(e.getMessage(), e);
     }
-    TemplateName template;
-    String templateName = args[0].trim().toUpperCase();
+  }
+
+  public static void main(String[] args) {
+    CommandLine cmd = parseArguments(args);
+    String templateNameString = cmd.getOptionValue(TEMPLATE_NAME);
+    Properties properties = cmd.getOptionProperties("prop");
+    String[] remainingArgs = cmd.getArgs();
+    LOGGER.info("Template name: {}", templateNameString);
+    LOGGER.info("Properties: {}", properties);
+    LOGGER.info("Remaining args: {}", (Object) remainingArgs);
+    PropertyUtil.registerProperties(properties);
+
+    TemplateName templateName;
     try {
-      template = TemplateName.valueOf(templateName);
+      templateName = TemplateName.valueOf(templateNameString.trim().toUpperCase());
     } catch (IllegalArgumentException ex) {
-      LOGGER.error("Unexpected template name :{}", templateName, ex);
-      throw ex;
+      throw new IllegalArgumentException(String.format("Unexpected template name: %s", templateNameString), ex);
     }
-    LOGGER.info("Running job for {}", template);
-    runSparkJob(template);
+    LOGGER.info("Running job for {}", templateName);
+    runSparkJob(templateName, cmd.getArgs());
   }
 
+  static final Map<TemplateName, Function<String[], BaseTemplate>> templateFactories =
+      ImmutableMap.<TemplateName, Function<String[], BaseTemplate>>builder()
+          .put(TemplateName.WORDCOUNT, (args) -> new WordCount())
+          .put(TemplateName.HIVETOGCS, (args) -> new HiveToGCS())
+          .put(TemplateName.HIVETOBIGQUERY, (args) -> new HiveToBigQuery())
+          .put(TemplateName.PUBSUBTOBQ, (args) -> new PubSubToBQ())
+          .put(TemplateName.GCSTOBIGQUERY, (args) -> new GCStoBigquery())
+          .put(TemplateName.S3TOBIGQUERY, (args) -> new S3ToBigQuery())
+          .put(TemplateName.SPANNERTOGCS, (args) -> new SpannerToGCS())
+          .put(TemplateName.GENERAL, GeneralTemplate::of)
+          .build();
```

Comment:
Using a map of template names to factories instead of a big switch statement. This also asserts that every template actually implements the BaseTemplate interface, which we didn't before.